### ISSUE TITLE
class inferred as abstract when it isn't.

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#8245]
Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBAhgRgYygSwgBzCALrOBnLEGBLCAUywAswATAKFEiiwE81kA7AcxXUxwAK4LGARgANnToJxMPHigARGFjkUA2gBUA%2BmIC6ACiFgRY8QEoAXHSi2oIAERPjpiVGCYlKtThly8AHRODlJ2UACSAHIAYgDy2gCqAEoAMpZQBCBQALxQkSbh6OJk5BxYZPRhjk4A6mRwsDQ0IGTyUGJlMJycPDSiAK6lqljIYBww4igcHiAQKqMcuGD9OFRkUDTeeBRBTqF2AALwmcSkFNSVdjRkwBtgAO4c4mAwNAbb4sDmUAC0AHx5MZkdIBUE2Q7HQinchUWjg2zXW7PV7vMifb7-KA6MQgsHwqAAYigEGQAA9OMTzrQ8PjEVMPKj0b8AfkOMD8WFkLcPsAAlE4olUigFPksIU0MUhhVrGFZXYiMhtoCxUUSmQyhUAKIgcAgAwOfIRGLxZIpKBocAAN2Q1xo7k8VEVGy2FAAhA5zBy7BbOFgDMAHABlMjrADePL5xsFKQAvh6pH42gBZFjKVTbP1pnzqX16Ky0m53R7It48jEsoEyuXmkC%2B-WKB5PF40HrxgtI5uMr7MqZYKtylpYfogRYAFgATFIU1mMwZvoSoIAZcigAGEYBwOCYpgR1yMVOtIUQSO1ZG0HNOXVgQkA)

```python
from abc import abstractmethod
from typing import Protocol

class Dataset[T_co](Protocol):
    r"""Protocol for Dataset class."""

    INFO_URL: str = NotImplemented
    r"""Web address containing documentational information about the dataset."""

    @abstractmethod
    def download(self) -> None: ...
    @abstractmethod
    def load(self) -> T_co: ...

    # mixin methods
    def info(self) -> None:
        if self.INFO_URL is NotImplemented:
            raise NotImplementedError("No INFO_URL provided for this dataset!")
        print(f"See {self.INFO_URL}")

class MyDataset(Dataset[int]):
    def download(self) -> None:
        print("Downloading")

    def load(self) -> int:
        return 42

MyDataset()  # ❌ Cannot instantiate abstract class "MyDataset"
```

It seems pyright infers that `info` is abstract due to the two lines:

```python
        if self.INFO_URL is NotImplemented:
            raise NotImplementedError("No INFO_URL provided for this dataset!")
```

if `NotImplementedError` is replaced with another type of Exception, `info` is no longer inferred as abstract.

**Comment by randolf-scholz** [comment]
Moreover, it seems this affects only Protocol types. If the base class is changes to `abc.ABC` or nothing, then it also disappears.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.370
